libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
This ref count is needed to keep the ibv_device memory regardless whether its kernel device was deleted. Cleaning the memory upon device deletion is wrong and may crash the application. 
Application can get the 'IBV_EVENT_DEVICE_FATAL' event to know about its device deletion and upon ibv_close_device the memory will be freed.

The rare race that the kernel device was deleted and was re-claimed by another device just before that ibv_open_device is called  is orthogonal to this series and might happen without this series as well.

This can be handled in some future support in kernel side or in the user area by using the timestamp creation just after calling cmd_fd = open(devpath, O_RDWR | O_CLOEXEC); as part of __ibv_open_device().